internal/client/consensus/grandpa/communication.go

```diff
@@ -0,0 +1,847 @@
+	topic := roundTopic[H, Hasher](round, setID)
+
+	nb.gossipEngineMtx.Lock()
+	messages := nb.gossipEngine.MessagesFor(topic)
+	nb.gossipEngineMtx.Unlock()
+
+	incoming := make(chan primitives.SignedMessage[H, N])
+	go func() {
+		defer close(incoming)
+		for notification := range messages {
+			var decoded gossipMessageVDT[H, N]
+			err := scale.Unmarshal(notification.Message, &decoded)
+			if err != nil {
+				logger.Debugf("Skipping malformed message %v: %v", notification, err)
+				continue
+			}
+			message, err := decoded.Value()
+			if err != nil {
+				logger.Debugf("Skipping malformed message %v: %v", notification, err)
+				continue
+			}
+			switch message := message.(type) {
+			case gossipMessageVote[H, N]:
+				// check signature.
+				if !voters.Contains(message.Message.ID) {
+					logger.Debugf("Skipping message from unknown voter %v", message.Message.ID)
+					continue
+				}
+				signedMessage := message.Message
+				incoming <- signedMessage
+			default:
+				logger.Debugf("Skipping unknown message type")
+				continue
+			}
+		}
+	}()
+
+	sender := make(chan primitives.SignedMessage[H, N])
+	outgoing := outgoingMessages[H, N, Hasher]{
+		keystore: ks,
+		round:    round,
+		setID:    setID,
+		network:  &nb.gossipEngine,
+		sender:   sender,
+		hasVoted: hasVoted,
+	}
+
+	// Combine incoming votes from external GRANDPA nodes with outgoing votes from our own GRANDPA voter to have a
+	// single vote-import-pipeline.
+	combinedIncoming := make(chan primitives.SignedMessage[H, N])
+	go func() {
+		defer close(combinedIncoming)
+		var incomingOk, senderOk bool
+		var msg primitives.SignedMessage[H, N]
+		for {
+			select {
+			case msg, incomingOk = <-incoming:
+				if !incomingOk {
+					continue
+				}
+				combinedIncoming <- msg
+			case msg, senderOk = <-sender:
+				if !senderOk {
+					continue
+				}
+				combinedIncoming <- msg
+			}
+			if !incomingOk && !senderOk {
+				break
+			}
+		}
+	}()
+
+	return combinedIncoming, outgoing
+}
+
+// Set up the global communication streams.
+func (nb *networkBridge[H, N, Hasher]) globalCommunication(
+	setID SetID,
+	voters *grandpa.VoterSet[primitives.AuthorityID],
+	isVoter bool,
+) (chan communicationIn[H, N], commitsOut[H, N, Hasher]) {
+	authorities := make([]primitives.AuthorityID, voters.Len())
+	for i, ivi := range voters.Voters() {
+		authorities[i] = ivi.ID
+	}
+	nb.validator.noteSet(setID, authorities, func(to []peerid.PeerID, msg neighborPacket[N]) {
+		select {
+		case nb.neighborSender <- peerIDsNeighborPacket[N]{PeerIDs: to, NeighborPacket: msg}:
+		default:
+			panic("wtf?")
+		}
+	})
+
+	topic := globalTopic[H, Hasher](setID)
+	incoming := incomingGlobal(&nb.gossipEngine, topic, voters, nb.validator, nb.neighborSender)
+
+	outgoing := newCommitsOut(&nb.gossipEngine, setID, isVoter, nb.validator, nb.neighborSender)
+
+	// 	let outgoing = outgoing.with(|out| {
+	// 		let voter::CommunicationOut::Commit(round, commit) = out;
+	// 		future::ok((round, commit))
+	// 	});
```

Comment:
is this comment need? maybe transform into an explanation?

Reply:
This is to remind myself in the upcoming PRs, that this is transforming the channel into a subset.  I'm not sure I need to do this, but it will be removed in future PRs.